File: src/api/communityConfigs.ts
import { invoke } from '@tauri-apps/api/core';

import type { CommentsPage, CommunityConfig, ConfigPerspective, ConfigType } from '../types';

export const COMMENTS_PAGE_SIZE = 40;

export function fetchCommunityConfigs(): Promise<CommunityConfig[]> {
  return invoke('fetch_community_configs');
}

export function fetchConfigComments(
configId : string,
options ?: { limit?: number; offset?: number; parentId?: string | null },
)        : Promise<CommentsPage> {
  return invoke('fetch_config_comments', {
    configId,
    limit   : options?.limit ?? COMMENTS_PAGE_SIZE,
    offset  : options?.offset ?? 0,
    parentId: options?.parentId ?? null,
  });
}

export function fetchConfigReaction(configId: string): Promise<1 | -1 | null> {
  return invoke('fetch_config_reaction', { configId });
}

export function setConfigReaction(configId: string, reaction: 1 | -1): Promise<void> {
  return invoke('set_config_reaction', { configId, reaction });
}

export function clearConfigReaction(configId: string): Promise<void> {
  return invoke('clear_config_reaction', { configId });
}

export function postConfigComment(configId: string, body: string, parentId?: string | null): Promise<void> {
  return invoke('post_config_comment', { configId, body, parentId: parentId ?? null });
}

export function updateConfigComment(commentId: string, body: string): Promise<void> {
  return invoke('update_config_comment', { commentId, body });
}

export function deleteConfigComment(commentId: string): Promise<void> {
  return invoke('delete_config_comment', { commentId });
}

export function createConfig(
name       : string,
note       : string,
type       : ConfigType,
perspective: ConfigPerspective | null,
data       : unknown,
)          : Promise<void> {
  return invoke('create_config', { name, note, type, perspective, data });
}

export function updateConfig(
configId   : string,
name       : string,
note       : string,
type       : ConfigType,
perspective: ConfigPerspective | null,
data       : unknown,
)          : Promise<void> {
  return invoke('update_config', { configId, name, note, type, perspective, data });
}

export function deleteConfig(configId: string): Promise<void> {
  return invoke('delete_config', { configId });
}

File: src-tauri/src/commands/community_configs.rs
use tauri::State;

use crate::application::{community_board, discord_auth_flow};
use crate::dto::{CommentsPageDto, CommunityConfigDto};
use crate::invoke_error::{InvokeErrorDto, invoke_err, sign_in_required};
use crate::state::AppState;

#[tauri::command]
pub async fn fetch_community_configs() -> Result<Vec<CommunityConfigDto>, InvokeErrorDto> {
    community_board::list_configs().await.map_err(invoke_err)
}

#[tauri::command]
pub async fn fetch_config_comments(
    config_id: String,
    limit    : Option<u32>,
    offset   : Option<u32>,
    parent_id: Option<String>,
) -> Result<CommentsPageDto, InvokeErrorDto> {
    community_board::list_comments(&config_id, limit, offset, parent_id.as_deref())
        .await
        .map_err(invoke_err)
}

#[tauri::command]
pub async fn fetch_config_reaction(state: State<'_, AppState>, config_id: String) -> Result<Option<i8>, InvokeErrorDto> {
    let session = discord_auth_flow::current_session(&state).ok_or_else(sign_in_required)?;
    let token   = discord_auth_flow::access_token(&state).await.map_err(invoke_err)?.ok_or_else(sign_in_required)?;
    community_board::my_reaction(&token, &session.user_id, &config_id).await.map_err(invoke_err)
}

#[tauri::command]
pub async fn set_config_reaction(state: State<'_, AppState>, config_id: String, reaction: i8) -> Result<(), InvokeErrorDto> {
    let session = discord_auth_flow::current_session(&state).ok_or_else(sign_in_required)?;
    let token   = discord_auth_flow::access_token(&state).await.map_err(invoke_err)?.ok_or_else(sign_in_required)?;
    community_board::set_reaction(&token, &session.user_id, &config_id, reaction)
        .await
        .map_err(invoke_err)
}

#[tauri::command]
pub async fn clear_config_reaction(state: State<'_, AppState>, config_id: String) -> Result<(), InvokeErrorDto> {
    let session = discord_auth_flow::current_session(&state).ok_or_else(sign_in_required)?;
    let token   = discord_auth_flow::access_token(&state).await.map_err(invoke_err)?.ok_or_else(sign_in_required)?;
    community_board::clear_reaction(&token, &session.user_id, &config_id).await.map_err(invoke_err)
}

#[tauri::command]
pub async fn post_config_comment(
    state: State<'_, AppState>,
    config_id: String,
    body     : String,
    parent_id: Option<String>,
) -> Result<(), InvokeErrorDto> {
    let token = discord_auth_flow::access_token(&state).await.map_err(invoke_err)?.ok_or_else(sign_in_required)?;
    community_board::post_comment(&token, &config_id, &body, parent_id.as_deref())
        .await
        .map_err(invoke_err)
}

#[tauri::command]
pub async fn update_config_comment(state: State<'_, AppState>, comment_id: String, body: String) -> Result<(), InvokeErrorDto> {
    let token = discord_auth_flow::access_token(&state).await.map_err(invoke_err)?.ok_or_else(sign_in_required)?;
    community_board::update_comment(&token, &comment_id, &body)
        .await
        .map_err(invoke_err)
}

#[tauri::command]
pub async fn delete_config_comment(state: State<'_, AppState>, comment_id: String) -> Result<(), InvokeErrorDto> {
    let token = discord_auth_flow::access_token(&state).await.map_err(invoke_err)?.ok_or_else(sign_in_required)?;
    community_board::delete_comment(&token, &comment_id).await.map_err(invoke_err)
}

#[tauri::command]
pub async fn create_config(
    state: State<'_, AppState>,
    name       : String,
    note       : String,
    r#type     : String,
    perspective: Option<String>,
    data       : serde_json::Value,
) -> Result<(), InvokeErrorDto> {
    let token = discord_auth_flow::access_token(&state).await.map_err(invoke_err)?.ok_or_else(sign_in_required)?;
    community_board::create_config(&token, &name, &note, &r#type, perspective.as_deref(), data).await.map_err(invoke_err)
}

#[tauri::command]
pub async fn update_config(
    state: State<'_, AppState>,
    config_id  : String,
    name       : String,
    note       : String,
    r#type     : String,
    perspective: Option<String>,
    data       : serde_json::Value,
) -> Result<(), InvokeErrorDto> {
    let token = discord_auth_flow::access_token(&state).await.map_err(invoke_err)?.ok_or_else(sign_in_required)?;
    community_board::update_config(&token, &config_id, &name, &note, &r#type, perspective.as_deref(), data)
        .await
        .map_err(invoke_err)
}

#[tauri::command]
pub async fn delete_config(state: State<'_, AppState>, config_id: String) -> Result<(), InvokeErrorDto> {
    let token = discord_auth_flow::access_token(&state).await.map_err(invoke_err)?.ok_or_else(sign_in_required)?;
    community_board::delete_config(&token, &config_id).await.map_err(invoke_err)
}